src/electionguard/async.hpp

```diff
@@ -0,0 +1,324 @@
+            virtual void call() = 0;
+            virtual ~Impl() {}
+        };
+        std::unique_ptr<Impl> _pimpl;
+        template <typename F> struct TypedImpl : Impl {
+            F callable;
+            TypedImpl(F &&callable) : callable(std::move(callable)) {}
+            void call() { callable(); }
+        };
+
+      public:
+        CallableWrapper() {}
+        template <typename F>
+        CallableWrapper(F &&callable) : _pimpl(new TypedImpl<F>(std::move(callable)))
+        {
+        }
+
+        CallableWrapper(CallableWrapper &&other) : _pimpl(std::move(other._pimpl)) {}
+
+        void operator()() { _pimpl->call(); }
+
+        CallableWrapper &operator=(CallableWrapper &&other)
+        {
+            _pimpl = std::move(other._pimpl);
+            return *this;
+        }
+
+        CallableWrapper(const CallableWrapper &) = delete;
+        CallableWrapper(CallableWrapper &) = delete;
+        CallableWrapper &operator=(const CallableWrapper &) = delete;
+
+        void call() { _pimpl->call(); }
+    };
+
+    /// <summary>
+    /// A simple thread pool that uses a common work queue for all threads
+    /// </summary>
+    class ThreadPool
+    {
+      public:
+        ThreadPool(const std::uint_fast32_t &thread_count = std::thread::hardware_concurrency())
+            : _thread_count(thread_count)
+        {
+            try {
+                for (unsigned i = 0; i < _thread_count; ++i) {
+                    _threads.push_back(std::thread(&ThreadPool::worker, this));
+                }
+            } catch (...) {
+                _running = false;
+                throw;
+            }
+        }
+
+        ThreadPool(const ThreadPool &other) = delete;
+        ThreadPool(const ThreadPool &&other) = delete;
+
+        ~ThreadPool()
+        {
+            wait();
+            _running = false;
+            for (auto &thread : _threads) {
+                thread.join();
+            }
+        }
+
+        ThreadPool &operator=(ThreadPool other) = delete;
+        ThreadPool &operator=(ThreadPool &&other) = delete;
+
+        template <typename F> std::future<typename std::result_of<F()>::type> submit(F callable)
+        {
+            typedef typename std::result_of<F()>::type result_type;
+
+            std::packaged_task<result_type()> task(std::move(callable));
+            std::future<result_type> result(task.get_future());
+            _taskQueue.push(std::move(task));
+            return result;
+        }
+
+      private:
+        std::uint_fast32_t _thread_count;
+        std::atomic<bool> _running = true;
+        AsyncQueue<CallableWrapper> _taskQueue;
+        std::mutex queue_mutex = {};
+        std::vector<std::thread> _threads;
+
+        void wait()
+        {
+            while (_running) {
+                if (_taskQueue.empty() == true) {
+                    break;
+                }
+                std::this_thread::yield();
+            }
+        }
+
+        void worker()
```

Comment:
i chose to use a single work queue here for simplicity but it is possible this leads to caching issues and may need more complexity in the future